subsys/net/lib/sockets/socketpair.c

```diff
@@ -0,0 +1,1087 @@
+
+#ifdef CONFIG_USERSPACE
+static inline int z_vrfy_zsock_socketpair(int family, int type, int proto,
+					  int *sv)
+{
+	int ret;
+	int tmp[2];
+
+	ret = z_impl_zsock_socketpair(family, type, proto, tmp);
+	Z_OOPS(z_user_to_copy(sv, tmp, sizeof(tmp));
+	return ret;
+}
+
+#include <syscalls/zsock_spair_mrsh.c>
+#endif /* CONFIG_USERSPACE */
+
+/**
+ * Write data to one end of a @ref spair
+ *
+ * Data written on one file descriptor of a socketpair can be read at the
+ * other end using common POSIX calls such as read(2) or recv(2).
+ *
+ * If the underlying file descriptor has the @ref O_NONBLOCK flag set then
+ * this function will return immediately. If no data was written on a
+ * non-blocking file descriptor, then -1 will be returned and @ref errno will
+ * be set to @ref EAGAIN.
+ *
+ * Blocking write operations occur when the @ref O_NONBLOCK flag is @em not
+ * set and there is insufficient space in the @em remote @ref spair.pipe.
+ *
+ * Such a blocking write will suspend execution of the current thread until
+ * one of two possible results is received on the @em remote
+ * @ref spair.read_signal:
+ *
+ * 1) @ref SPAIR_SIG_DATA - data has been read from the @em remote
+ *    @ref spair.pipe. Thus, allowing more data to be written.
+ *
+ * 2) @ref SPAIR_SIG_CANCEL - the @em remote socketpair endpoint was closed
+ *    Receipt of this result is analagous to SIGPIPE from POSIX
+ *    ("Write on a pipe with no one to read it."). In this case, the function
+ *    will return -1 and set @ref errno to @ref EINTR.
+ *
+ * @param obj the address of an @ref spair object cast to `void *`
+ * @param buffer the buffer to write
+ * @param count the number of bytes to write from @p buffer
+ *
+ * @return on success, a number > 0 representing the number of bytes written
+ * @return -1 on error, with @ref errno set appropriately.
+ */
+static ssize_t spair_write(void *obj, const void *buffer, size_t count)
+{
+	ssize_t res;
+
+	bool is_connected;
+	size_t avail;
+	bool is_nonblock;
+	bool will_block;
+	size_t bytes_written;
+
+	bool have_local_sem = false;
+	bool have_remote_sem = false;
+	struct spair *const spair = (struct spair *)obj;
+	struct spair *remote = NULL;
+
+	if (obj == NULL || buffer == NULL || count == 0) {
+		NET_DBG("invalid argument");
+		errno = EINVAL;
+		res = -1;
+		goto out;
+	}
+
+	res = k_sem_take(&spair->sem, K_FOREVER);
+	if (res < 0) {
+		NET_ERR("failed to take local sem: %d", res);
+		errno = -res;
+		res = -1;
+		goto out;
+	}
+
+	have_local_sem = true;
+
+	remote = z_get_fd_obj(spair->remote,
+			(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+	is_connected = sock_is_connected(spair);
+	avail = is_connected ? spair_write_avail(spair) : 0;
+	is_nonblock = sock_is_nonblock(spair);
+	will_block = (count > avail) && !is_nonblock;
+
+	NET_DBG("count: %u is_connected: %u avail: %u is_nonblock: %u "
+		"will_block: %u", count, is_connected, avail, is_nonblock,
+		will_block);
+
+	if (!is_connected) {
+		errno = EPIPE;
```

Comment:
Aha, so it's EPIPE after all, docstring is stale?

Reply:
Yes. Fixed.

Reply:
Will push shortly. Apologies, but it takes a while to make code changes, then split all of them into separate code and test commits. I'll just leave it as one commit until it's been fully reviewed this time.

Reply:
> Apologies, but it takes a while to make code changes, then split all of them into separate code and test commits. I'll just leave it as one commit until it's been fully reviewed this time.

I don't understand this. You're familiar with `git add -p`, `git commit -m f`, `git stash save`, `git rebase -i master`, `git stash pop`, right? Aka, you don't lump up all stuff thogether, and then split it up, you keep it finely split up and squash ("fix-up") together as needed (usually, incrementally using a cycle above, it's very easy to reason about 5-line-or-less changes).